extensions/wikia/Chat2/Chat.class.php

```diff
@@ -72,435 +82,207 @@ public static function echoCookies() {
 	 *
 	 * @return bool true on success, returns an error message as a string on failure.
 	 */
-	public static function promoteChatModerator( $userNameToPromote, $promotingUser ) {
-		ChatHelper::info( __METHOD__ . ': Method called', [
-			'userNameToPromote' => $userNameToPromote,
-			'promotingUser' => $promotingUser
+	public static function promoteModerator( $subjectUserName, $adminUser ) {
+		self::info( __METHOD__ . ': Method called', [
+			'userNameToPromote' => $subjectUserName,
+			'promotingUser' => $adminUser
 		] );
-		$CHAT_MOD_GROUP = 'chatmoderator';
 
-		$userToPromote = User::newFromName( $userNameToPromote );
+		$subjectUser = User::newFromName( $subjectUserName );
 
-		if ( !( $userToPromote instanceof User ) ) {
-			$errorMsg = wfMessage( 'chat-err-invalid-username-chatmod', $userNameToPromote )->text();
+		if ( !( $subjectUser instanceof User ) ) {
+			$errorMsg = wfMessage( 'chat-err-invalid-username-chatmod', $subjectUserName )->text();
 
 			return $errorMsg;
 		}
 
 		// Check if the userToPromote is already in the chatmoderator group.
 		$errorMsg = '';
-		if ( in_array( $CHAT_MOD_GROUP, $userToPromote->getEffectiveGroups() ) ) {
-			$errorMsg = wfMessage( "chat-err-already-chatmod", $userNameToPromote, $CHAT_MOD_GROUP )->text();
+		if ( in_array( self::CHAT_MODERATOR, $subjectUser->getEffectiveGroups() ) ) {
+			$errorMsg = wfMessage( "chat-err-already-chatmod", $subjectUserName, self::CHAT_MODERATOR )->text();
+		} else if ( self::canPromoteModerator( $subjectUser, $adminUser ) ) {
+			self::doPromoteModerator( $adminUser, $subjectUser );
 		} else {
-			$changeableGroups = $promotingUser->changeableGroups();
-			$promotingUserName = $promotingUser->getName();
-			$isSelf = ( $userToPromote->getName() == $promotingUserName );
-			$addableGroups = array_merge( $changeableGroups['add'], $isSelf ? $changeableGroups['add-self'] : [] );
-
-			if ( in_array( $CHAT_MOD_GROUP, $addableGroups ) ) {
-				// Adding the group is allowed. Add the group, clear the cache, run necessary hooks, and log the change.
-				$oldGroups = $userToPromote->getGroups();
-
-				self::permissionsService()->addToGroup( $promotingUser, $userToPromote, $CHAT_MOD_GROUP );
-
-				if ( $userToPromote instanceof User ) {
-					$removegroups = [];
-					$addgroups = [ $CHAT_MOD_GROUP ];
-					wfRunHooks( 'UserRights', [ &$userToPromote, $addgroups, $removegroups ] );
-				}
-
-				// Update user-rights log.
-				$newGroups = array_merge( $oldGroups, [ $CHAT_MOD_GROUP ] );
-
-				// Log the rights-change.
-				Chat::addLogEntry( $userToPromote, $promotingUser, [
-					Chat::makeGroupNameListForLog( $oldGroups ),
-					Chat::makeGroupNameListForLog( $newGroups )
-				], 'chatmoderator' );
-			} else {
-				$errorMsg = wfMessage( "chat-err-no-permission-to-add-chatmod", $CHAT_MOD_GROUP )->text();
-			}
+			$errorMsg = wfMessage( "chat-err-no-permission-to-add-chatmod", self::CHAT_MODERATOR )->text();
 		}
 
 		return ( $errorMsg == "" ? true : $errorMsg );
```

Comment:
i don't like functions which are returning mixed types, what do you think about returning always `$errorMsg` string and check for emptiness while processing result?